cvat/apps/dataset_manager/bindings.py

```diff
@@ -2151,10 +2151,66 @@ def reduce_fn(acc, v):
                 raise CvatImportError("Image {}: can't import annotation "
                     "#{} ({}): {}".format(item.id, idx, ann.type.name, e)) from e
 
-    for track in tracks.values():
-        track['elements'] = list(track['elements'].values())
-        instance_data.add_track(instance_data.Track(**track))
+    def _validate_track_shapes(shapes):
+        new_shapes = []
+        prev_shape = None
+        # infer the keyframe shapes and keep only them
+        for shape in shapes:
+            prev_is_visible = prev_shape and not prev_shape.outside
+            cur_is_visible = shape and not shape.outside
+            if not cur_is_visible:
+                if prev_is_visible:
+                    shape = shape._replace(keyframe=True)
```

Comment:
Actually, it seems like we need to check the distance after the prev shape here as well, like it's done below. Because it's possible that the current invisible shape has a gap after the last visible shape, and we need to insert an outside keyframe after the last visible shape in this case.